pkg/api/rest/rest.go

```diff
@@ -202,20 +202,19 @@ type Redirector interface {
 	ResourceLocation(ctx api.Context, id string) (remoteLocation *url.URL, transport http.RoundTripper, err error)
 }
 
-// ConnectHandler is a handler for HTTP connection requests. It extends the standard
-// http.Handler interface by adding a method that returns an error object if an error
-// occurred during the handling of the request.
-type ConnectHandler interface {
-	http.Handler
-
-	// RequestError returns an error if one occurred during handling of an HTTP request
-	RequestError() error
+// Responder provides standard response behavior.
+type Responder interface {
+	// Object writes the provided object to the response. Invoking this method multiple times is undefined.
+	Object(statusCode int, obj runtime.Object)
+	// Error writes the provided error to the response. This method may only be invoked once.
+	Error(err error)
 }
 
 // Connecter is a storage object that responds to a connection request
 type Connecter interface {
-	// Connect returns a ConnectHandler that will handle the request/response for a request
-	Connect(ctx api.Context, id string, options runtime.Object) (ConnectHandler, error)
+	// Connect returns an http.Handler that will handle the request/response for a given API invocation.
+	// The provided responder may be used for common API responses.
+	Connect(ctx api.Context, id string, options runtime.Object, r Responder) (http.Handler, error)
```

Comment:
Hrm. That opens the possibility that the http.Handler would mutate the
ResponseWriter with a wrapper.  Do we want to enable that?

On Mon, Oct 12, 2015 at 6:56 PM, Daniel Smith <notifications@github.com>
wrote:

> In pkg/api/rest/rest.go
> <https://github.com/kubernetes/kubernetes/pull/15442#discussion_r41810054>
> :
>
> >  }
> >
> >  // Connecter is a storage object that responds to a connection request
> >  type Connecter interface {
> > -	// Connect returns a ConnectHandler that will handle the request/response for a request
> > -	Connect(ctx api.Context, id string, options runtime.Object) (ConnectHandler, error)
> > +	// Connect returns an http.Handler that will handle the request/response for a given API invocation.
> > +	// The provided responder may be used for common API responses.
> > +	Connect(ctx api.Context, id string, options runtime.Object, r Responder) (http.Handler, error)
>
> I see-- the caller needs to have parameterized the responder. Wouldn't it
> be much clearer to just have Responder's methods take an http.Response? And
> possibly save every caller the work.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/15442/files#r41810054>.
>